### DISCUSSION TITLE
[Feature request]:Error Compiling TOPAS v3.9 and TOPAS-nBio v2.0 on CentOS 7.9.2009 Cluster

### DISCUSSION THREAD

**Discussion by WKjie** [topas-nbio/TOPAS-nBio#58] [Feature-Request] [answered]
### Feature request

Hello,

I am currently working on installing TOPAS v3.9 and TOPAS-nBio v2.0 on a SLURM-managed cluster running CentOS 7.9.2009. I successfully installed TOPAS v3.9, but I am encountering issues when trying to install TOPAS-nBio v2.0.

I chose these versions based on the old installation guides supporting CentOS 7, but I am uncertain about the correct GCC and CMake versions required. Additionally, I would like to know if the latest versions of TOPAS and TOPAS-nBio can be installed on this cluster (CentOS 7.9.2009).

In my attempts to install TOPAS-nBio v2.0, I tried various GCC versions through module loading, but failed during the 'make' process, encountering the following error:

libextensions.a(TsEmDNAChemistry.cc.o): In function `G4TDNAOneStepThermalizationModel<DNA::Penetration::Meesungnoen2002>::Initialise(G4ParticleDefinition const*, G4DataVector const&)':
TsEmDNAChemistry.cc:(.text._ZN32G4TDNAOneStepThermalizationModelIN3DNA11Penetration15Meesungnoen2002EE10InitialiseEPK20G4ParticleDefinitionRK12G4DataVector[_ZN32G4TDNAOneStepThermalizationModelIN3DNA11Penetration15Meesungnoen2002EE10InitialiseEPK20G4ParticleDefinitionRK12G4DataVector]+0x36d): undefined reference to `G4Exception(char const*, char const*, G4ExceptionSeverity, std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >&)'
libextensions.a(DrDNAMoleculeEncounterStepper.cc.o): In function `DrDNAMoleculeEncounterStepper::CalculateMinTimeStep(double, double)':
DrDNAMoleculeEncounterStepper.cc:(.text+0x1965): undefined reference to `G4Exception(char const*, char const*, G4ExceptionSeverity, std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >&)'
libextensions.a(DrDNAMoleculeEncounterStepper.cc.o): In function `G4ITFinder<G4Molecule>::FindNearest(G4Molecule const*, int)':
DrDNAMoleculeEncounterStepper.cc:(.text._ZN10G4ITFinderI10G4MoleculeE11FindNearestEPKS0_i[_ZN10G4ITFinderI10G4MoleculeE11FindNearestEPKS0_i]+0x379): undefined reference to `G4Exception(char const*, char const*, G4ExceptionSeverity, std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >&)'
libextensions.a(TsDNAFirstOrderReaction.cc.o): In function `TsDNAFirstOrderReaction::SetReaction(G4MolecularConfiguration const*, double)':
TsDNAFirstOrderReaction.cc:(.text+0x871): undefined reference to `G4Exception(char const*, char const*, G4ExceptionSeverity, std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >&)'
libextensions.a(TsDNAFirstOrderReaction.cc.o): In function `TsDNAFirstOrderReaction::SetReaction(G4MolecularConfiguration const*, std::vector<G4MolecularConfiguration*, std::allocator<G4MolecularConfiguration*> >, double)':
TsDNAFirstOrderReaction.cc:(.text+0xb08): undefined reference to `G4Exception(char const*, char const*, G4ExceptionSeverity, std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >&)'
libextensions.a(TsDNARemoveInMaterial.cc.o):TsDNARemoveInMaterial.cc:(.text+0x3a8): more undefined references to `G4Exception(char const*, char const*, G4ExceptionSeverity, std::__cxx11::basic_ostringstream<char, std::char_traits<char>, std::allocator<char> >&)' follow
collect2: error: ld returned 1 exit status
make[2]: *** [bin/topas] Error 1
make[1]: *** [CMakeFiles/topas.dir/all] Error 2
make: *** [all] Error 2

I would appreciate any guidance or recommendations regarding:

The appropriate versions of GCC and CMake needed to install TOPAS-nBio v2.0 alongside TOPAS v3.9 on CentOS 7.
Whether the latest versions of TOPAS and TOPAS-nBio can be installed on CentOS 7.9.2009.
Any specific configuration settings or dependencies that I might be missing.
Potential issues with CentOS supporting these versions of TOPAS and TOPAS-nBio that I should be aware of.

Thank you in advance for your assistance!

### TOPAS-nBio category

Geometry

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Attestation

- [X] I testify that I have fully read the bug reporting guideline beforehand.

**Comment by tmasilela** [comment, marked as answer]
Hi,

I'm not too sure about the old version, but I did successfully install the newest version of TOPAS (OpenTOPAS v4.0.0) and TOPAS-nBio (v3.0) on a RHEL cluster recently. My guess is that you should be able to install it on a CentOS cluster in a similar fashion but you could encounter one or two problems along the way which you'll need to troubleshoot.

For the installation on the cluster I use I largely followed the [Debian quickStart guide](https://github.com/OpenTOPAS/OpenTOPAS/blob/main/OpenTOPAS_quickStart_Debian.md). I started from Step 5 since the cluster already had all the command-line tools that you are prompted to install in the previous steps. And since I didn't care about visualization on the cluster I changed the cmake line in Step 7 to:

`cmake ../geant4-v11.1.3 -DGEANT4_INSTALL_DATA=OFF -DGEANT4_BUILD_MULTITHREADED=ON -DCMAKE_INSTALL_PREFIX=../geant4-install`

This avoids any potential problems arising with qt@5. Then assuming you have already downloaded TOPAS-nBio, the cmake line in Step 8.4 becomes:

`cmake ../OpenTOPAS -DCMAKE_INSTALL_PREFIX=../OpenTOPAS-install \
           -DTOPAS_EXTENSIONS_DIR=../TOPAS-nBio`

Hope that helps.

All the best,
Thongchai
